### ISSUE TITLE
Ender 3 dual extruder setup not working

### ISSUE THREAD

**Issue by B-rad888** [Ultimaker/Cura#8222]
**Application version**
4.6.1

**Platform**
Windows 10

**Printer**
Custom Printer and/or Ender 3 Pro

**Reproduction steps**

Trying to set up dual extruders Part 1 with Ender 3 Pro Printer:
1. Cura wont allow setting Ender 3 pro to 2 extruders.
2. Selecting drop down for extruders gives no option to change number of extruders

Trying to set up dual extruders part 2 with custom printer:
1. Cura allows selecting multiple extruders.
2. Export ender 3 profiles from ender 3 printer in cura
2. Import Ender 3 pro profiles into custom printer gives error:

 Failed to import profile from C:/Users/****/Desktop/Cura Profile export/Standard Quality.curaprofile: Could not find a quality type standard for the current configuration.

**Actual results**
No way to get Ender profiles with dual extruders.

**Expected results**
Ender 3 printer should allow me to select number of extruders. 

**Additional information**
I really want to get this working please help.

**Comment by B-rad888** [comment]
Now I'm getting really irritated! I just copied all the settings from the ender 3 profile to the custom printer. I made profiles for each quality (  0.12mm - Super Quality, 0.16mm - Dynamic Quality, 0.28mm - Low Quality, 0.2mm - Standard Quality). After making these profiles one by one after about 3 hours of comparison, some settings are not saved. This is ridiculous! I just want to select 2 extruders.

**Comment by Ghostkeeper** [comment]
Indeed the authors of the profile never added any second extruder. Cura doesn't know that it's possible. And if it is, it doesn't know the nozzle offsets, where to go to switch extruders, or how to switch. We don't have enough information to select 2 extruders, so we can't. To make this work, you'd need to add additional [extruder definitions](https://github.com/Ultimaker/Cura/tree/master/resources/extruders) to Cura. We can't add these extruder definitions ourselves, since we don't have every printer on the market at the Ultimaker office, so we're really just dependent on community contributions to maintain those profiles. Do you have the expertise to create an extruder definition like that?

> Failed to import profile from C:/Users/****/Desktop/Cura Profile export/Standard Quality.curaprofile: Could not find a quality type standard for the current configuration.

The export of the profile doesn't work since the profile is made for a quality level which is not available on the Custom FFF Printer. The profile also only contains the changes you made on the default, but in this case the default it's based on is not available for the other printer.

**Comment by B-rad888** [comment]
Thank you for some clarification. For now I'm just trying to add a 2 in 1 out hotend:
https://amzn.to/3jbJ1I1
So there is no offset besides the z offset.

 Do you have the expertise to create an extruder definition like that?

I'm trying to figure it out. There is no guides that I can find on the matter, but I see lots of people who have done it. It wasn't to hard to add an extra extruder to marlin. Setting up the profile in Cura was a pain. I think I got that part finally saved. 

I believe the next step is to define tool changing (retraction and exruding) for the y merger in the nozzle.  I understand there is complications for adding dual extrusion to every printer. There are may variants. However the lack of guidance on this one is a little overwhelming.

**Comment by nallath** [comment]
You can find information about how to add machines to Cura here: https://github.com/Ultimaker/Cura/wiki/Adding-new-machine-profiles-to-Cura

There are alos other articles on the wiki about some of these internal structures.

**Comment by B-rad888** [comment]
@nallath Thanks for the info. I'm not sure if I would have to create a new printer. I was able to use the custom printer and manually create profiles just like the Ender 3. I copied a lot of the code from the CR-X for the tool changing. I tweaked a bunch of perimeters. I'm still not there yet. I started a few prints with it. So far I can get extruder 0 to print the first layer, and the beginning of a purge tower. After the tool change to extruder 1 the nozzle comes back and starts scraping the bed. I'm not sure whats going on. Its been 2 days to get to this point. I feel like I'm close, but it's not right. I don't know if adding a machine would fix this. I'm also unsure how to create those files. Its not clear. Do I just copy and paste into notepad++ ? Still I'm not sure if starting over and creating those files would give a different result.

**Comment by Ghostkeeper** [comment]
> After the tool change to extruder 1 the nozzle comes back and starts scraping the bed. I'm not sure whats going on.

If you got this far you're indeed close. It's probably going to the [end position](https://github.com/Ultimaker/Cura/blob/6f3298385f96170c6507fc0abc3e97d844ce4f77/resources/definitions/fdmextruder.def.json#L146) of the extruder, where it needs to switch. I suggest you set the `machine_extruder_end_pos_abs` and `machine_extruder_start_pos_abs` to `False` to make it switch in a position relative to the current nozzle position where it is when the switch is instructed. You can then move 10mm aside or so to hopefully move away from the outer wall to prevent a blob there (so `machine_extruder_end_pos_x` to 10). Or just leave the position at 0,0, which means, don't move when changing extruder.
I believe the Printer Settings plug-in from the Marketplace exposes those settings so it's pretty easy to optimize them using that plug-in.

**Comment by B-rad888** [comment]
Thank you for the advice! I will try changing the values to false. I ended up creating the new machine profiles from scratch from the templates. It should be easier to upgrade to newer versions of Cura now. I'm almost ready to slice a new model and try it. Is there a way to get the Creality quality and profiles to work with the new machine profiles? I tried almost everything to get it to work. I even created a second creality_base (creawesome base printer) to link the new machine definition file and number of extruder definition files. This didn't quite work. It would be nice to have only files to drag and drop into cura for each update. I see Cura 4.7 is out. recreating the print quality profiles for each new update will get tiresome eventually.

**Comment by Ghostkeeper** [comment]
Adding this to the `metadata` section should make it select the Creality set of quality profiles:

    "quality_definition": "creality_base",

You can also use this (in the root of the JSON) to make it start from the setting values in the definition of the Ender 3:

    "inherits": "creality_ender3"

But you'll always need to give it its own extruders and variants.

**Comment by B-rad888** [comment]
Thanks for the tips. Unfortunately, it doesn't work. The profiles say unsupported, and the nozzles say empty. So It's not really usable. Maybe I'm doing it wrong? I have also tried
`"inherits": "creality base"`
Same thing. They show up but no quality profiles or nozzles to select or use.
I'm inserting the following in metadata:
`        "machine_extruder_trains": {
            "0": "2in1_Extruder_1",
            "1": "2in1_Extruder_2"`
I can only get it to work with:
`"inherits": "fdmprinter",`
However anything related to Creawesome will not work. I've tried about everything I can think of.

**Comment by Ghostkeeper** [comment]
You'll still need to copy the extruders and variants (you'll find them in the `extruders` folder and the `variants` folder. The "variants" are really just the nozzle profiles, the name is a bit of legacy.

**Comment by B-rad888** [comment]
@Ghostkeeper Thank you very much for your help! I got it working mostly. It turns out that my original issue was hardware and not software. I installed dual Z axis with a belt. The belt was to tight and caused skipping. Hence why it would print a layer, raise, and skip then come back lower. I got my profiles all set up though finally. :) I found quite a few issues/bugs with dual extrusion in Cura after all that work. I think I will have to make a new thread for those though. Thanks again for your help!
